metadata-ingestion/src/datahub/ingestion/source/tableau/tableau.py

```diff
@@ -3052,33 +3067,84 @@ def _get_ownership(self, user: str) -> Optional[OwnershipClass]:
         return None
 
     def emit_project_containers(self) -> Iterable[MetadataWorkUnit]:
-        for _id, project in self.tableau_project_registry.items():
-            parent_container_key: Optional[ContainerKey] = None
-            if project.parent_id:
-                parent_container_key = self.gen_project_key(project.parent_id)
-            elif self.config.add_site_container and self.site and self.site.id:
-                parent_container_key = self.gen_site_key(self.site.id)
+        visited_project_ids: Set[str] = set()
+
+        def emit_project_in_topological_order(
+            project_: TableauProject,
+        ) -> Iterable[MetadataWorkUnit]:
+            """
+            The source_helpers.py file includes a function called auto_browse_path_v2 that automatically generates
+            BrowsePathsV2. This auto-generation relies on the correct sequencing of MetadataWorkUnit containers.
+            The emit_project_in_topological_order function ensures that parent containers are emitted before their child
+            containers.
+            This is a recursive function.
+            """
+            if project_.parent_id is None:
+                project_key = self.gen_project_key(project_.id)
+
+                if project_key.guid() in visited_project_ids:
+                    return
+
+                visited_project_ids.add(project_key.guid())
+
+                parent_container_key: Optional[ContainerKey] = None
+                # set site as parent container
+                if self.config.add_site_container and self.site and self.site.id:
+                    parent_container_key = self.gen_site_key(self.site.id)
```

Comment:
why is this after the `visited_project_ids` check?

also, why are we adding to `visited_project_ids` in two places?

Reply:
reorganized the code